### PR TITLE
[FIXED JENKINS-31185] Added hg support

### PR DESCRIPTION
**Pull request by matthewcmorgan**
First pass attempt to add hg support.  
Tested working on privately hosted Jenkins v. 1.635.  
Only supports web hook payload, not post service payload, though that wouldn't be hard to add.

Resolves [Jenkins-31185](https://issues.jenkins-ci.org/browse/JENKINS-31185).

### CODE REVIEW

**Comment by Xarthisius** [comment]
@fbelzunc is there anything blocking this pull request?

**Comment by fbelzunc** [comment]
Hello, sorry I didn't have time to review this. I will do tomorrow and provide my feedback.

File: src/main/java/com/cloudbees/jenkins/plugins/BitbucketJobProbe.java [hunk 1]
```diff
@@ -54,7 +54,7 @@ public void triggerMatchingJobs(String user, String url, String scm) {
                             } else LOGGER.log(Level.FINE, "{0} SCM doesn't match remote repo {1}", new Object[]{job.getName(), remote});
                         }
                     } else
-                        LOGGER.log(Level.FINE, "{0} hasn't BitBucketTrigger set", job.getName());
+                    LOGGER.log(Level.FINE, "{0} hasn't BitBucketTrigger set", job.getName());
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketJobProbe.java]: This line shouldn't be touched as there is not any modification.

File: src/main/java/com/cloudbees/jenkins/plugins/BitbucketJobProbe.java [hunk 2]
```diff
@@ -85,10 +84,32 @@ private boolean match(SCM scm, URIish url) {
                         return true;
                 }
             }
+        } else if (scm instanceof MercurialSCM) {
+            try {
+                URI hgUri = new URI(((MercurialSCM) scm).getSource());
+                String remote = url.toString();
+                if (almostMatches(hgUri, remote))
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketJobProbe.java]: Please, after an `if` open/close brackets `{`, `}` even if there is only one line.

File: src/main/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessor.java [hunk 3]
```diff
@@ -1,9 +1,6 @@
 package com.cloudbees.jenkins.plugins;
-
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessor.java]: You don't need to touch anything here.

File: src/main/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessor.java [hunk 4]
```diff
@@ -1,9 +1,6 @@
...
 import java.util.logging.Logger;
-
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessor.java]: You don't need to touch anything here.

File: src/main/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessor.java [hunk 5]
```diff
@@ -1,9 +1,6 @@
...
 import javax.servlet.http.HttpServletRequest;
-
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessor.java]: You don't need to touch anything here.

File: src/main/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessor.java [hunk 6]
```diff
@@ -31,55 +28,26 @@ public void processPayload(JSONObject payload, HttpServletRequest request) {
...
     private void processWebhookPayload(JSONObject payload) {
-        JSONObject repo = payload.getJSONObject("repository");
-        LOGGER.info("Received commit hook notification for "+repo);
-
-        String user = payload.getJSONObject("actor").getString("username");
-        String url = repo.getJSONObject("links").getJSONObject("html").getString("href");
-        String scm = repo.has("scm") ? repo.getString("scm") : "git";
+        if (payload.has("repository")) {
+            JSONObject repo = payload.getJSONObject("repository");
+            LOGGER.info("Received commit hook notification for "+repo);
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessor.java]: Please, at least change:

```
LOGGER.info("Received commit hook notification for "+repo);
```

per

```
LOGGER.info("Received commit hook notification for "+ repo);
```
Although the right way is:

```
LOGGER.log(Level.INFO, "Received commit hook notification for {0}", repo);
```

File: src/main/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessor.java [hunk 7]
```diff
@@ -31,55 +28,26 @@ public void processPayload(JSONObject payload, HttpServletRequest request) {
...
+
+            String user = payload.getJSONObject("actor").getString("username");
+            String url = repo.getJSONObject("links").getJSONObject("html").getString("href");
+            String scm = repo.has("scm") ? repo.getString("scm") : "git";
+
+            probe.triggerMatchingJobs(user, url, scm);
+        } else if (payload.has("scm")) {
+            LOGGER.info("Received commit hook notification for hg " + payload);
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessor.java]: Same structure needs to be done here:

```
LOGGER.log(Level.INFO, "Received commit hook notification for hg {0}", payload);
```

File: src/test/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessorTest.java [hunk 8]
```diff
@@ -43,9 +43,19 @@ public void testProcessWebhookPayload() {
...
+        JSONObject hgLoad = new JSONObject()
+            .element("scm", "hg")
+            .element("owner", new JSONObject()
+                .element("username", user))
+            .element("links", new JSONObject()
+                .element("html", new JSONObject()
+                    .element("href", url)));
+
         payloadProcessor.processPayload(payload, request);
-
```
Comment (fbelzunc) [inline src/test/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessorTest.java]: You don't need to change this.

File: src/main/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessor.java [hunk 9]
```diff
@@ -31,55 +28,26 @@ public void processPayload(JSONObject payload, HttpServletRequest request) {
...
-        "is_private": true,
-        "name": "Project X",
-        "owner": "marcus",
-        "scm": "git",
-        "slug": "project-x",
-        "website": "https://atlassian.com/"
-    },
-    "user": "marcus"
-}
-*/
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessor.java]: You don't need to change this.

**Comment by fbelzunc** [comment]
:ant: You only need to do small modifications. 

I would also suggest you change the title to the pull request to *[FIXED JENKINS-31185] Added hg support*. I believe that if you don't do that then Jira will not auto-find the Jira issue related to this pull request.

Thanks for contributing to BitBucket plugin!

**Comment by matthewcmorgan** [comment]
Sorry for the sloppy pull request, I will make these fixes and resubmit soon.  Thank you for taking the time to review.

**Comment by iNikem** [comment]
@matthewcmorgan friendly ping :) Do you plan to finish this or we can take it over? We are quite interested in the Mercurial support and were thinking about implementing it ourselves.

**Comment by matthewcmorgan** [comment]
I plan on cleaning this up over the holidays, should be resubmitted soon.  :)

**Comment by matthewcmorgan** [comment]
Okay, resubmitted with the fixes.  One line I couldn't undo as it wasn't complaining that I removed a newline but that I removed unnecessary tabs on the line.  If it matters I suppose we can use the online github editor to put them back in.  Everything else was cleaned up.  Let me know if anything else is needed.  Thank you for the review @fbelzunc

File: src/main/java/com/cloudbees/jenkins/plugins/BitbucketJobProbe.java [hunk 10]
```diff
@@ -81,14 +83,38 @@ private boolean match(SCM scm, URIish url) {
...
+                    if (GitStatus.looselyMatches(urIish, url)) {
                         return true;
+                    }
+                }
+            }
+        } else if (scm instanceof MercurialSCM) {
+            try {
+                URI hgUri = new URI(((MercurialSCM) scm).getSource());
+                String remote = url.toString();
+                if (almostMatches(hgUri, remote)) {
```
Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketJobProbe.java]: You should use [MercurialStatus.looselyMatches](https://github.com/jenkinsci/mercurial-plugin/blob/17a01f2097e3a90568b42705c9a0d223695bdef5/src/main/java/hudson/plugins/mercurial/MercurialStatus.java#L61) instead.

Comment (matthewcmorgan) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketJobProbe.java]: Yeah, I saw that when I was developing this, and I remember I couldn't use it.  I tried implementing it again and realized why I included the almostMatches method...  The 
[MercurialStatus.looselyMatches](https://github.com/jenkinsci/mercurial-plugin/blob/17a01f2097e3a90568b42705c9a0d223695bdef5/src/main/java/hudson/plugins/mercurial/MercurialStatus.java#L61) method is not public unfortunately.

Comment (fbelzunc) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketJobProbe.java]: You are right.

Comment (matthewcmorgan) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketJobProbe.java]: Sure thing!

**Comment by fbelzunc** [comment]
Thank you!

**Pull request merged**